image/cas/interface.go

```diff
@@ -0,0 +1,39 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cas implements generic content-addressable storage.
+package cas
+
+import (
+	"io"
+)
+
+// Engine represents a content-addressable storage engine.
+type Engine interface {
+
+	// Put adds a new blob to the store.
+	Put(writer io.Reader) (digest string, err error)
```

Comment:
On Mon, Jul 04, 2016 at 01:44:34AM -0700, Simone Gotti wrote:
> > +	// Put adds a new blob to the store.
> > +	Put(writer io.Reader) (digest string, err error)
> 
> Since there isn't mediaType information provided in put and returned
> by get, should the user autodetect the mediaType or is this just not
> needed since it'll be already known because the user will always
> fetch a blob starting from a parent referencing object (descriptor,
> manifest, manifest list) that already provides the mediaType?

I don't think it's needed for consuming code.  But we might need this
information for a Docker-registry-based backend, because where you
push the data depends on what you're pushing [1].

> This assumes that all the objects will be part of a tree and there's
> always the way to retrieve the mediaType by the parent.

It doesn't necessarily mean all blobs will be part of a tree, it just
assumes that if you know you want blob $DIGEST, you already know it's
type.  In a world where all Merkle links are descriptors [2], that's a
safe assumption.

> Instead if there isn't this assumption then a mediaType should be
> returned or it should be detected by the user for every
> object. (another example can be a List method that can return blobs
> by mediaTypes)

I don't think list-CAS-blobs-by-media-type is an API we need to
support.  In oci-store, you only use ListBlobs (which filters by
media-type) for garbage collection, and you don't need that if you
drill down from refs.

[1]: https://docs.docker.com/registry/spec/api/#/pushing-an-image
     Layers to blobs/uploads/, manifests to manifests/<reference>, …
[2]: https://github.com/opencontainers/image-spec/blame/ca58cdfc7c8200c6c62287e677fadb0074cc29cc/descriptor.md#L11